### ISSUE TITLE
Inconsitent type compatibility when for a type with a call signature and and index signature

### ISSUE THREAD

**Issue by dragomirtitian** [Microsoft/TypeScript#23226]
**TypeScript Version:** 2.9.0-dev

**Search Terms:** index signature intersection

**Code**

```ts
type INOk = {
    (): string;
    [name: string] : number;
}

type IOk = {
    (): string;
} & {
    [name: string] : number;
}

declare let val:(() => "") & { foo: number; }
let ok: IOk = val; // This works
let nok: INOk = val; // This does not

```

**Expected behavior:**
Both `IOk` and `INOk` have the same public structure, they both have a call signature and are indexable, both assign statements should be valid.

**Actual behavior:**
The second assign statement fails with the message `Index signature is missing in type '(() => "") & { foo: number; }'.`

**Playground Link:** [link](https://www.typescriptlang.org/play/index.html#src=type%20INOk%20%3D%20%7B%0A%20%20%20%20()%3A%20string%3B%0A%20%20%20%20%5Bname%3A%20string%5D%20%3A%20number%3B%0A%7D%0A%0Atype%20IOk%20%3D%20%7B%0A%20%20%20%20()%3A%20string%3B%0A%7D%20%26%20%7B%0A%20%20%20%20%5Bname%3A%20string%5D%20%3A%20number%3B%0A%7D%0A%0Adeclare%20let%20val%3A(()%20%3D%3E%20%22%22)%20%26%20%7B%20foo%3A%20number%3B%20%7D%0Alet%20ok%3A%20IOk%20%3D%20val%3B%20%2F%2F%20This%20works%0Alet%20nok%3A%20INOk%20%3D%20val%3B%20%2F%2F%20This%20does%20not%0A)
**Related Issues:** https://github.com/Microsoft/TypeScript/issues/15300

**Notes**
Looking at the code, it appears that for `IOk` type compatibility is checked for each constituent of the intersection type, so we will have :

```ts
isRelatedTo (typeof val, IOk) = 
      isRealtedTo(typeof val, () => ""))   // == True, since val has a call signature
             ( // Above equal to 
                   isRelatedTo(() => "", () => "") // ==  True
                   ||
                   isRelatedTo({ foo: number; }, () => "") // == False but it's does not matter
             )
      &&
      isRealtedTo(typeof val, { [name: string] : number }) // = True since the { foo: number} part of typeof val has an inferable index (ie isObjectTypeWithInferableIndex( { foo: number} ) will return true)
            ( // Above equal to 
                   isRelatedTo(() => "",  { [name: string] : number }) // ==  False but does not matter
                   ||
                   isRelatedTo({ foo: number; },  { [name: string] : number }) // == True, 
             )

```

While for `INOk` the relation is checked directly, since `INOk` can't be split into constituents and we have 
```ts
isRelatedTo (typeof val, INOk) = 
              isRelatedTo(() => "",  INOk) // ==  False INOk has index, but ()=> "" does not
              ||
              isRelatedTo({ foo: number; },  INOk) // == False, No compatible call signature
```
So then the checker falls back to structural checking (`recursiveTypeRelatedTo`) but this fails as well because when checking for index compatibility (inside `indexTypesRelatedTo`), it decides that the `typeof value` (`(() => "") & { foo: number; }`)  does not have an inferable index (`isObjectTypeWithInferableIndex`  returns false because the intersection type does not have a `symbol` and even if it did the condition for inferable index checks that the type does not have a call signature (`!typeHasCallOrConstructSignatures(type)`))

**Comment by dragomirtitian** [comment]
Proposed solution, allow `isObjectTypeWithInferableIndex` to return true for intersection types, if any of the intersection constituents have an inferable index.

https://github.com/dragomirtitian/TypeScript4ExtJS/commit/1d7723ad58a36bd45833de3ae89db01bdf8d0263

**Comment by RyanCavanaugh** [comment]
@dragomirtitian we discussed this and think that the intersection type behavior (the assignment is OK) is the desired behavior.

Before we change anything, though, we'd like to understand how you found this - what does the unreduced repro look like?

**Comment by dragomirtitian** [comment]
@RyanCavanaugh This was posted as a question on [stackoverflow](https://stackoverflow.com/questions/49672942/typescript-implement-an-interface-with-both-call-signature-and-indexing-signatu/49673416#49673416) , the OP did not provide more information about their  use-case.

**Comment by RyanCavanaugh** [comment]
Just to expand on what's going on here

One rule is that a type `T` is assignable to the type `X & Y` if `T` is assignable to `X` and `T` is assignable to `Y`; this is one of the first principles of intersection types. There's no direct reasoning about `X & Y` during this operation.

Another rule is that a type `T` is assignable to a type `{ [s: string]: U }` if all the declared properties of `T` are assignable to `U` *and* `T` has no call or construct signatures. The second clause is there because a type like `{ [s: string]: Foo }` is a common "map" type and we don't want bare functions (which have no properties at all) to be assignable to these maps simply because they don't have properties.

This leads us to a "consistency triangle" problem. We think the first rule is correct, and we think the second rule is correct, but we also think `X & Y` should generally behave the same as its "normalized" form, but doesn't.

A proposed change is that types with *at least one property* are not subject to the "... does not have call or construct signatures" rule. This would be reasonably straightforward and square the circle, so to speak, but we're not really going to make a change unless we first understand how someone noticed this in the first place. Sometimes people go out hunting for inconsistencies and this hunt usually turns up *something*; we'd rather spend our risk budget on "real" issues if possible.

TL;DR if anyone noticed this in real code please tell us what it looked like so we can understand the severity/priority of it.

**Comment by RaduSzasz** [comment]
@RyanCavanaugh I posted the StackOverflow question @dragomirtitian is referring to. I didn't encounter the issue in the real world; I am doing my Master Thesis on a subject related to TypeScript and tried to get a better understanding of what assignments does the type system allow.

Thank you for your detailed comment.

**Comment by RyanCavanaugh** [comment]
>  I would argue that the root cause here is that nothing should be implicitly assignable to the common "map" type unless it is itself a compatible map type

This indeed used to be the behavior, but it was changed due to feedback. We got a lot of "bug" reports because people wrote code like this
```ts
function checkAges(map: { [name: string]: number }) { ... }
const x = {
  "bob": 32,
  "alice": 26,
  "eve": 42
};

// Error, wat, of course this should be valid!
checkAges(x);
```
We got enough reports that we seeded a StackOverflow question for it https://stackoverflow.com/questions/22077023/why-cant-i-indirectly-return-an-object-literal-to-satisfy-an-index-signature-re . The old behavior seems right on paper, but in practice led to a lot of unnecessary friction.
